lumicks/pylake/tests/data/mock_confocal.py

```diff
@@ -186,31 +214,347 @@ def make_slice(data):
+    return np.arange(min(2, image.ndim)) if axes is None else axes
 
 
 def generate_scan(
-    name, scan_data, pixel_sizes_nm, start=4, dt=7, samples_per_pixel=5, line_padding=3
+    name,
+    image,
+    pixel_sizes_nm,
+    axes=None,
+    start=20 * 10**9,
+    dt=635 * 10**8,
+    samples_per_pixel=5,
+    line_padding=3,
+    multi_color=False,
 ):
+    return _generate_confocal(
+        name,
+        Scan,
+        image,
+        multi_color=multi_color,
+        pixel_sizes_nm=pixel_sizes_nm,
+        axes=_scan_axes(image, axes),
+        start=start,
+        dt=dt,
+        samples_per_pixel=samples_per_pixel,
+        line_padding=line_padding,
+    )
+
+
+def _generate_confocal(
+    name,
+    confocal_class,
+    image,
+    multi_color,
+    pixel_sizes_nm,
+    axes,
+    start,
+    dt,
+    samples_per_pixel,
+    line_padding,
+):
+    start = np.int64(start)
+    dt = np.int64(dt)
+
     confocal_file, metadata, stop = MockConfocalFile.from_image(
-        np.swapaxes(scan_data, -1, -2),
+        image,
         pixel_sizes_nm=pixel_sizes_nm,
-        axes=np.arange(scan_data.ndim),
-        start=np.int64(start),
-        dt=np.int64(dt),
+        axes=axes,
+        start=start,
+        dt=dt,
         samples_per_pixel=samples_per_pixel,
         line_padding=line_padding,
+        multi_color=multi_color,
     )
 
-    return Scan(name, confocal_file, start, stop, metadata)
+    return confocal_class(name, confocal_file, start, stop, metadata)
+
+
+def generate_kymo_with_ref(
+    name,
+    image,
+    pixel_size_nm=10.0,
+    start=20 * 10**9,
+    dt=625 * 10**8,
+    samples_per_pixel=5,
+    line_padding=3,
+):
+    kymo = generate_kymo(name, image, pixel_size_nm, start, dt, samples_per_pixel, line_padding)
+    ref = _generate_confocal_reference(
+        Kymo,
+        kymo.file,
+        kymo._metadata,
+        image,
+        multi_color=image.ndim > 2,
+        axes=[0],
+        start=start,
+        dt=dt,
+        samples_per_pixel=samples_per_pixel,
+        line_padding=line_padding,
+    )
+    return kymo, ref
+
+
+def generate_scan_with_ref(
+    name,
+    image,
+    pixel_sizes_nm,
+    axes=None,
+    start=20 * 10**9,
+    dt=625 * 10**8,
+    samples_per_pixel=5,
+    line_padding=3,
+    multi_color=False,
+):
+    scan = generate_scan(
+        name, image, pixel_sizes_nm, axes, start, dt, samples_per_pixel, line_padding, multi_color
+    )
+    ref = _generate_confocal_reference(
+        Scan,
+        scan.file,
+        scan._metadata,
+        image,
+        multi_color=multi_color,
+        axes=_scan_axes(image, axes),
+        start=start,
+        dt=dt,
+        samples_per_pixel=samples_per_pixel,
+        line_padding=line_padding,
+    )
+    return scan, ref
+
+
+def _generate_confocal_reference(
+    confocal_class,
+    confocal_file,
+    metadata,
+    image,
+    multi_color,
+    axes,
+    start,
+    dt,
+    samples_per_pixel,
+    line_padding,
+):
```

Comment:
I wonder, for separation of concerns, should we split from here down in a separate file like `mock_confocal_reference`? Just an idea as, personally, I find really long files difficult to work with, especially when you're constantly scrolling up and down to see the web of calls

Reply:
I would like to keep the generation of scans/kymos and the corresponding reference data together, to emphasize that they should be treated as a "unit" and belong together. As stated above, in the long run, we probably could get rid of the separation of the functions `generate_scan()`/`generate_kymo()` and `generate_scan_with_ref()`/`generate_kymo_with_ref()` and thereby reduce number of lines.